File: game/scripts/vscripts/pve/abilities/heroes/queen/queen_attack/queen_attack.lua
queen_attack = class({})
LinkLuaModifier("modifier_queen_attack", "pve/abilities/heroes/queen/queen_attack/modifier_queen_attack", LUA_MODIFIER_MOTION_NONE)

function queen_attack:GetCastAnimationCustom()		return ACT_DOTA_CAST_ABILITY_1 end
function queen_attack:GetPlaybackRateOverride()    return 0.9 end
function queen_attack:GetCastPointSpeed() 			return 0 end

function queen_attack:OnSpellStart()
    local point = self:GetCursorPosition()
    self:LaunchProjectile(point, 1500, self:GetCastRange(Vector(0,0,0), nil), 20)
    EmitSoundOn("Hero_QueenOfPain.ShadowStrike", self:GetCaster())
end

function queen_attack:LaunchProjectile(vPoint, nSpeed, nDistance, nDamage)
    local caster = self:GetCaster()
    local origin = caster:GetOrigin()
    local point = vPoint
    local projectile_direction = Direction2D(origin, point)
    local poison_duration = 4.0
    local fading_slow_duration = 4.0
    local fading_slow_pct = 80

    CustomEntitiesLegacy:ProjectileAttack(caster, {
        tProjectile = {
            EffectName = "particles/queen/queen_dagger.vpcf",
            vSpawnOrigin = origin + Vector(0, 0, 96),
            fDistance = nDistance,
            fStartRadius = 70,
            Source = caster,
            vVelocity = projectile_direction * nSpeed,
            UnitBehavior = PROJECTILES_DESTROY,
            TreeBehavior = PROJECTILES_NOTHING,
            WallBehavior = PROJECTILES_DESTROY,
            GroundBehavior = PROJECTILES_NOTHING,
            fGroundOffset = 0,
            UnitTest = function(_self, unit) return unit:GetUnitName() ~= "npc_dummy_unit" and not CustomEntitiesLegacy:Allies(_self.Source, unit) end,
            OnUnitHit = function(_self, unit) 
                local damage_table = {
                    victim = unit,
                    attacker = _self.Source,
                    damage = nDamage,
                    damage_type = DAMAGE_TYPE_PHYSICAL,
                    ability = self
                }
                ApplyDamage(damage_table)

                unit:AddNewModifier(_self.Source, self, "modifier_queen_attack", { duration = poison_duration })
                unit:AddNewModifier(_self.Source, self, "modifier_generic_fading_slow", { 
                    duration = fading_slow_duration,
                    max_slow_pct = fading_slow_pct 
                })
            end,
            OnFinish = function(_self, pos)
                EFX("particles/units/heroes/hero_queenofpain/queen_shadow_strike_explosion.vpcf", PATTACH_WORLDORIGIN, nil, {
                    cp0 = pos,
                    cp3 = pos,
                })
                EmitSoundOn('Hero_QueenOfPain.ShadowStrike.Target', caster)
            end,
        }
    })
end

if IsClient() then require("wrappers/abilities") end
Abilities.Castpoint(queen_attack)

File: src/vscripts/pve/abilities/heroes/queen/queen_attack.ts
import { CustomAbility } from "../../../../abilities/framework/custom_ability";
import { CustomModifier } from "../../../../abilities/framework/custom_modifier";
import { registerAbility, registerModifier } from "../../../../lib/dota_ts_adapter";
import { ModifierFadingSlow } from "../../../../modifiers/modifier_fading_slow";
import { areUnitsAllied, direction2D } from "../../../../util";

@registerAbility("queen_attack")
export class QueenAttack extends CustomAbility {
    GetAnimation() {
        return GameActivity.DOTA_CAST_ABILITY_1;
    }

    GetPlaybackRateOverride() {
        return 0.9;
    }

    GetCastingCrawl() {
        return 0;
    }

    OnSpellStart() {
        const point = this.GetCursorPosition();
        const distance = this.GetCastRange(Vector(0, 0, 0), undefined);
        this.LaunchProjectile(point, 1500, distance, 20);
        EmitSoundOn("Hero_QueenOfPain.ShadowStrike", this.caster);
    }

    LaunchProjectile(point: Vector, speed: number, distance: number, damage: number) {
        const origin = this.caster.GetOrigin();
        const projectileDirection = direction2D(origin, point);
        const poisonDuration = 4.0;
        const fadingSlowDuration = 4.0;
        const fadingSlowPct = 80;

        this.ProjectileAttack({
            source: this.caster,
            attackType: "basic",
            effectName: "particles/queen/queen_dagger.vpcf",
            distance: distance,
            startRadius: 70,
            spawnOrigin: origin.__add(Vector(0, 0, 96)),
            velocity: projectileDirection.__mul(speed),
            groundOffset: 0,
            unitTest: (unit, projectile) => !areUnitsAllied(projectile.getSource(), unit),
            onUnitHit: (unit, projectile) => {
                ApplyDamage({
                    victim: unit,
                    attacker: projectile.getSource(),
                    damage: damage,
                    damage_type: DamageTypes.PHYSICAL,
                    ability: this
                });

                ModifierQueenAttack.apply(unit, projectile.getSource(), this, { duration: poisonDuration });
                ModifierFadingSlow.apply(unit, projectile.getSource(), undefined, {
                    duration: fadingSlowDuration,
                    maxSlowPct: fadingSlowPct
                });
            },
            onFinish: projectile => {
                EFX(
                    "particles/units/heroes/hero_queenofpain/queen_shadow_strike_explosion.vpcf",
                    ParticleAttachment.WORLDORIGIN,
                    undefined,
                    {
                        cp0: projectile.getPosition(),
                        cp3: projectile.getPosition()
                    }
                );
                EmitSoundOn("Hero_QueenOfPain.ShadowStrike.Target", this.caster);
            }
        });
    }
}

@registerModifier("modifier_queen_attack")
class ModifierQueenAttack extends CustomModifier {
    OnCreated() {
        if (IsServer()) {
            this.StartIntervalThink(1.0);
        }
    }

    OnIntervalThink() {
        if (IsServer()) {
            ApplyDamage({
                attacker: this.caster,
                damage: 5,
                damage_type: DamageTypes.PURE,
                victim: this.parent
            });
        }
    }

    GetEffectName() {
        return "particles/generic_gameplay/rune_haste.vpcf";
    }

    GetEffectAttachType() {
        return ParticleAttachment.OVERHEAD_FOLLOW;
    }

    GetStatusEffectName() {
        return "particles/status_fx/status_effect_terrorblade_reflection.vpcf";
    }
}
